### PR TITLE
Authenticate and wrap HTTP request headers with context supplier

### PR DESCRIPTION
**Pull request by albertzaharovits**
Builds on top of https://github.com/elastic/elasticsearch/pull/92220 .

Adds authentication to HTTP headers, and packs the authenticated context to be used when executing the URL handler.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java [hunk 1]
```diff
@@ -249,6 +189,46 @@ public Exception getInboundException() {
         return inboundException;
     }
 
+    protected io.netty.handler.codec.http.HttpRequest getNettyRequest() {
+        return request;
+    }
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:194]: I need in order to get to the netty `HttpHeaders` that contains the authenticated context.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java [hunk 2]
```diff
@@ -249,6 +189,46 @@ public Exception getInboundException() {
...
+        if (httpMethod == HttpMethod.CONNECT) {
+            return RestRequest.Method.CONNECT;
+        }
+
+        throw new IllegalArgumentException("Unexpected http method: " + httpMethod);
+    }
+
+    protected static Map<String, List<String>> wrapHttpHeaders(HttpHeaders httpHeaders) {
+        return new HttpHeadersMap(httpHeaders);
+    }
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:230]: I need to reuse `translateRequestMethod` and `wrapHttpHeaders` when building the `BasicHttpRequest` in the new channel handler.
This way the `HttpRequest` instances, which are built for the complete request (including the body contents), and the `BasicHttpRequest` instances, which are built for the body-less requests from the new channel handler, reuse the same code.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java [hunk 3]
```diff
@@ -368,30 +402,52 @@ protected void initChannel(Channel ch) throws Exception {
...
+                        return headersAuthenticator.wrapNewMessage(httpRequest);
+                    }
+
+                    @Override
+                    protected HttpMessage createInvalidMessage() {
+                        DefaultHttpRequest httpRequest = (DefaultHttpRequest) super.createInvalidMessage();
+                        return headersAuthenticator.wrapNewMessage(httpRequest);
+                    }
+                };
+            } else {
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java:424]: This is intercepting the netty `HttpMessage` contruction during http request decoding, in order to wrap the default messages with a custom type that in addition carries the authentication context.
The authentication context is carried from the new channel inbound handler, through the netty pipeline, and the ES core code handling HTTP/1.1. pipelining, to the point where the thread context is prepared just before the request is dispatched to the handler.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java [hunk 4]
```diff
@@ -115,7 +120,12 @@ public Map<String, Supplier<HttpServerTransport>> getHttpTransports(
                 tracer,
                 TLSConfig.noTLS(),
                 null
-            )
+            ) {
+                @Override
+                protected void populateRequestThreadContext(RestRequest restRequest, ThreadContext threadContext) {
+                    dispatcherContext.accept(restRequest.getHttpRequest(), threadContext);
+                }
+            }
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java:128]: This is the Http handler when security is disabled.

In this case, just use the consumer that populates the request-wise thread context in the default way. That consumer is furnished by the `RestController`. This is unfortunate, it would've been simpler if the `Netty4HttpServerTransport` were responsilbe to populate the thread context before delegating to the REST dispatcher.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        );
+    }
+
+    public void authenticateMessage(HttpRequest request, Channel channel, ActionListener<Void> listener) {
+        assert request.headers() instanceof HttpHeadersAuthenticator.HttpHeadersWithAuthenticationContext;
+        final BasicHttpRequest requestToAuthenticate = wrapToBasicHttpRequest(request);
+        final Supplier<ThreadContext.StoredContext> emptyContext = threadContext.wrapRestorable(threadContext.newStoredContext());
+        final ActionListener<Void> contextPreservingListener = new ContextPreservingActionListener<>(emptyContext, listener);
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            populateThreadContext.apply(requestToAuthenticate, channel, threadContext);
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:76]: Populate the thread context before authentication, exactly the same way it is currently populated for "REST" requests.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            authenticate.accept(requestToAuthenticate, ActionListener.wrap(ignored -> {
+                final ThreadContext.StoredContext authenticatedContext = threadContext.newStoredContext();
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:78]: Following a successful authentication the thread context gets into an authenticated state.
This captures the full authenticated thread context, to be reinstanted downstream, when executing the  http handler.

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/part-3

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    private static HttpHeadersWithAuthenticationContext unwrapHeadersAuthenticationContext(BasicHttpRequest request) {
+        if (request instanceof Netty4HttpRequest == false) {
+            return null;
+        }
+        if (((Netty4HttpRequest) request).getNettyRequest().headers() instanceof HttpHeadersWithAuthenticationContext == false) {
+            return null;
+        }
+        return ((HttpHeadersWithAuthenticationContext) (((Netty4HttpRequest) request).getNettyRequest().headers()));
+    }
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:126]: The methods above are used to extract the attached authentication context from the request (they are used just before the time when the request is to be handled, and the authentication context needs to be instanted).
Unfortunately it relies on `instance of` checks (2 of them), only one of which can be "fixed" because the other uses netty interfaces.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                ctx.channel().eventLoop().submit(() -> validationSuccess(ctx));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                final HeaderValidationException toForward;
+                if (e instanceof HeaderValidationException) {
+                    toForward = (HeaderValidationException) e;
+                } else if (e instanceof ElasticsearchException) {
+                    toForward = new HeaderValidationException((ElasticsearchException) e, false);
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:102]: Changed this to, by default not close, the connection.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java [hunk 9]
```diff
@@ -41,59 +41,36 @@ public class Netty4HttpRequest implements HttpRequest {
 
     private final FullHttpRequest request;
     private final BytesReference content;
-    private final HttpHeadersMap headers;
+    private final Map<String, List<String>> headers;
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:44]: Changed in order for the `Netty4HttpRequest` to (continue to ) implement the `HttpRequest` which extends the new `BasicHttpRequest` interface.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 3 above)
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java:424]: This is the first of two changes to the netty pipeline.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java [hunk 10]
```diff
@@ -368,30 +402,52 @@ protected void initChannel(Channel ch) throws Exception {
...
+                    handlingSettings.maxHeaderSize(),
+                    handlingSettings.maxChunkSize()
+                );
+            }
             decoder.setCumulator(ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+            ch.pipeline().addLast("decoder", decoder);
+            if (headersAuthenticator != null) {
+                Netty4HttpHeaderValidator validator = new Netty4HttpHeaderValidator(headersAuthenticator::authenticateMessage);
+                ch.pipeline().addLast("header_validator", validator);
+            }
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java:436]: This is the second and last change to the pipeline. This is the channel handler that runs the authentication.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java (hunk 4 above)
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java:128]: I think this is one area where we can discuss improvements.

File: server/src/main/java/org/elasticsearch/common/network/NetworkModule.java [hunk 11]
```diff
@@ -122,6 +125,7 @@ public NetworkModule(
         NamedXContentRegistry xContentRegistry,
         NetworkService networkService,
         HttpServerTransport.Dispatcher dispatcher,
+        BiConsumer<BasicHttpRequest, ThreadContext> dispatcherContext,
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/common/network/NetworkModule.java:128]: We can choose to merge the `dispatcherContext` in the `HttpServerTransport.Dispatcher` interface, if we opt to go with presented approach where the `RestController` furnishes the thread context supplier, but it's the http server transport that employs it, because the Security plugin can alter the latter but not the former.

File: server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java [hunk 12]
```diff
@@ -376,11 +377,28 @@ void dispatchRequest(final RestRequest restRequest, final RestChannel channel, f
             if (badRequestCause != null) {
                 dispatcher.dispatchBadRequest(channel, threadContext, badRequestCause);
             } else {
+                populateRequestThreadContext0(restRequest, channel, threadContext);
                 dispatcher.dispatchRequest(restRequest, channel, threadContext);
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java:381]: This is the code that clears and scopes the thread context for a particular http request to be handled.
This is now extended to also populate the said thread context (with selected HTTP header values).

File: server/src/main/java/org/elasticsearch/http/BasicHttpRequest.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.elasticsearch.http;
+
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+import java.util.Map;
+
+public interface BasicHttpRequest {
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/http/BasicHttpRequest.java:16]: This interface is the common ground between the two request types, with or without a body.
It is now used in the `AuthenticationService` as well as the `AudittingService`.

File: server/src/main/java/org/elasticsearch/http/BasicHttpRequest.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Returns the HTTP method used in the HTTP request.
+     *
+     * @return the {@link RestRequest.Method} used in the REST request
+     * @throws IllegalArgumentException if the HTTP method is invalid
+     */
+    RestRequest.Method method();
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/http/BasicHttpRequest.java:24]: `Method` doesn't have to be a `RestRequest` inner enum, and can be moved here, but such a refactoring has a huge diff.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 15]
```diff
@@ -1592,42 +1605,82 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
...
+                }
             } else {
                 sslConfiguration = null;
+                populateClientCertificate = (channel, threadContext) -> {};
             }
+            final TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext = (httpRequest, channel, threadContext) -> {
+                dispatcherContext.accept(httpRequest, threadContext);
+                populateClientCertificate.accept(channel, threadContext);
+                RemoteHostHeader.process(channel, threadContext);
+            };
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1621]: The thread context before running the authentication in the channel handler. It reuses that "dispatch" context that's a sort of baseline (contains selected HTTP header values).

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java (hunk 4 above)
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java:128]: > But the Security pugin can only intercept the http server transport and not the rest controller, and it's the rest controller that contains the information necessary to populate the context (the list of HTTP headers to copy in the thread context).

For more details see the `Security` class, eg https://github.com/elastic/elasticsearch/pull/93656#discussion_r1111783159 .

Security needs to use the baseline thread context supplier, that's been built by the `RestController`, in the authenticating netty channel handler, which is constructed together with the http server transport by the Security plugin.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SSLEngineUtils.java [hunk 16]
```diff
@@ -29,45 +29,48 @@ public class SSLEngineUtils {
...
-        SSLEngine sslEngine = getSSLEngine(httpChannel);
-        extract(logger, threadContext, sslEngine, httpChannel);
+    public static void extractClientCertificates(Logger logger, ThreadContext threadContext, TcpChannel tcpChannel) {
+        extractClientCertificates(logger, threadContext, getNettyChannel(tcpChannel));
     }
 
-    public static void extractClientCertificates(Logger logger, ThreadContext threadContext, TcpChannel tcpChannel) {
-        SSLEngine sslEngine = getSSLEngine(tcpChannel);
-        extract(logger, threadContext, sslEngine, tcpChannel);
+    public static void extractClientCertificates(Logger logger, ThreadContext threadContext, Channel channel) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SSLEngineUtils.java:36]: Extracting the client certificates must work even when a `HttpChannel` is not available (it only needs the neety channel anyway).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java [hunk 17]
```diff
@@ -93,22 +89,17 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
             return;
         }
 
-        if (extractClientCertificate) {
-            HttpChannel httpChannel = request.getHttpChannel();
-            SSLEngineUtils.extractClientCertificates(logger, threadContext, httpChannel);
-        }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:99]: The Security plugin now incorporates this when it builds the thread context before running the authentication.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java [hunk 18]
```diff
@@ -93,22 +89,17 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
...
-                logger.trace("Authenticated REST request [{}] as {}", request.uri(), authentication);
+                logger.trace("Authenticated REST request [{}] as {}", httpRequest.uri(), authentication);
             }
-            secondaryAuthenticator.authenticateAndAttachToContext(request, ActionListener.wrap(secondaryAuthentication -> {
+            secondaryAuthenticator.authenticateAndAttachToContext(httpRequest, ActionListener.wrap(secondaryAuthentication -> {
                 if (secondaryAuthentication != null) {
-                    logger.trace("Found secondary authentication {} in REST request [{}]", secondaryAuthentication, request.uri());
+                    logger.trace("Found secondary authentication {} in REST request [{}]", secondaryAuthentication, httpRequest.uri());
                 }
-                RemoteHostHeader.process(request, threadContext);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:111]: This has also been moved to the consolidated thread context populator, **before** running authentication (here it is "after", but moving it "before" allows more audit events to benefit, and aids the transion from `RestRequest` to `BasicHttpRequest` in the auditing service).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java [hunk 19]
```diff
@@ -124,21 +115,28 @@ private void doHandleRequest(RestRequest request, RestChannel channel, NodeClien
     }
 
     protected void handleException(ActionType actionType, RestRequest request, RestChannel channel, Exception e) {
-        logger.debug(() -> format("%s failed for REST request [%s]", actionType, request.uri()), e);
         threadContext.sanitizeHeaders();
-        final RestStatus restStatus = ExceptionsHelper.status(e);
+        handleException(actionType, channel, e);
+    }
+
+    public static void handleAuthenticationFailed(RestChannel channel, Exception e) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:123]: "Exports" the authn failure response building logic to be used in the channel handler.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java [hunk 20]
```diff
@@ -240,7 +241,10 @@ private Authentication lookForExistingAuthentication(Authenticator.Context conte
             logger.error(() -> format("caught exception while trying to read authentication from request [%s]", context.getRequest()), e);
             throw context.getRequest().tamperedRequest();
         }
-        if (authentication != null && context.getRequest() instanceof AuthenticationService.AuditableRestRequest) {
+        // this throws if a new http request looks to be already authenticated
+        if (authentication != null
+            && (context.getRequest() instanceof AuthenticationService.AuditableHttpRequest)
+            && (((AuthenticationService.AuditableHttpRequest) context.getRequest()).request instanceof HttpRequest == false)) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java:247]: Right now, `HttpRequest`s are checked for authentication in two places: the original `SecurityRestFilter` and the new  `HttpHeadersAuthenticator`, so this check here must account for this fact. But it's likely that the final PR drops the `SecurityRestFilter`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java [hunk 21]
```diff
@@ -364,21 +364,21 @@ public String toString() {
...
             ThreadContext threadContext,
-            RestRequest request
+            BasicHttpRequest request
         ) {
             super(auditTrail, failureHandler, threadContext);
             this.request = request;
-            // There should never be an existing audit-id when processing a rest request.
-            this.requestId = AuditUtil.generateRequestId(threadContext);
+            // There might be an existing audit-id (e.g. generated by the rest request) but there might not be (e.g. an internal action)
+            this.requestId = AuditUtil.getOrGenerateRequestId(threadContext);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java:381]: Using `AuditUtil#getOrGenerateRequestId` instead of `AuditUtil#generateRequestId` is just a temporary measure.  Ultimately `HttpRequest`s will only be intercepted once so we can fall back to the stricter generate.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 22]
```diff
@@ -1592,42 +1605,82 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
...
+                public void dispatchBadRequest(RestChannel channel, ThreadContext threadContext, Throwable cause) {
+                    Exception authenticationException = HttpHeadersAuthenticator.extractAuthenticationException(
+                        channel.request().getHttpRequest()
+                    );
+                    if (authenticationException != null) {
+                        SecurityRestFilter.handleAuthenticationFailed(channel, authenticationException);
+                    } else {
+                        dispatcher.dispatchBadRequest(channel, threadContext, cause);
+                    }
+                }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1650]: Unfortunately the dispatcher needs to have the the "bad request" branch overridden, because of the small differences between the 400 errors and authn failure errors.

File: server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java (hunk 12 above)
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java:381]: Is this happening before or after authentication ?

**Review by jakelandis — COMMENTED** [review]
I'm still working though this review , but here are a couple early comments. Also, it would be helpful to have some high level explaination of how this works. I think I understand most of it ...but the order of execution between netty functions and what is in thread context is hard to follow.

File: server/src/main/java/org/elasticsearch/rest/RestRequestFilter.java [hunk 23]
```diff
@@ -28,50 +34,185 @@
 public interface RestRequestFilter {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/rest/RestRequestFilter.java:34]: why does this class need to change ?

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.TriConsumer;
+
+import java.util.ArrayDeque;
+
+public class Netty4HttpHeaderValidator extends ChannelInboundHandlerAdapter {
+
+    private final TriConsumer<HttpRequest, Channel, ActionListener<Void>> validator;
+    private ArrayDeque<HttpObject> pending = new ArrayDeque<>(4);
+    private STATE state = STATE.WAITING_TO_START;
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:34]: It is one instance per channel, given how it is instantiated in `Netty4HttpServerTransport`, with `ch.pipeline().addLast(..., new Validator())`.
There's always one thread per channel. This is core to netty's design and cannot be changed (ever).
So, no it's not required to be thread-safe.

[the remark this answers is not in the document]
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:34]: To be pedantic, there are multiple channels per thread, but for a given channel there's always the same single thread that calls the channel handlers.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.ReferenceCountUtil;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.TriConsumer;
+
+import java.util.ArrayDeque;
+
+public class Netty4HttpHeaderValidator extends ChannelInboundHandlerAdapter {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:30]: Of course we'll polish this, with javadoc, comments, etc. It's TimB creation from https://github.com/elastic/elasticsearch/pull/92220 . It also does slightly more than what we strictly need (closing a connection in case authn fails), and there is a small thing that I need to clarify myself when we forward the request but drop the contents.

All that's to say is that this needs trimmings, and it will be a separate PR for it to review and merge.

I propose we sync to explain how it works/supposed to work before we get to the polishing stage.

File: server/src/main/java/org/elasticsearch/rest/RestRequestFilter.java (hunk 23 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/RestRequestFilter.java:34]: Good question!
Short answer is that this filtering, which is needed for auditing, currently works on `RestRequest` types, but we now need it to work on HTTP request bodies.

And that's because auditing works with HTTP requests, because authentication now has to work with HTTP requests, and the two are very much entangled.

BUT I have a proposal to decouple auditing for authentication success from the actual authentication (which sounds bad, I admit) and this is the main reason: today, when we only support authentication on the "REST" channel (no transport client), and the authentication happens before the request body is available, and we want to audit the request body, we cannot cleanly couple authentication and auditing.

Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/RestRequestFilter.java:34]: Basically, it has to do with auditing, which is not properly addressed by this PR anyway (authn success is audited but not the request body), and for which I have a different proposal anyway (that will not require such changes).

File: server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java (hunk 12 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java:381]: In this PR this is happening **after** authentication (on `main` it happens before authentication).

The `populateRequestThreadContext` hook method is used to instante the authentication context, for the authn that happened earlier.

File: server/src/main/java/org/elasticsearch/http/HttpRequest.java [hunk 26]
```diff
@@ -11,51 +11,23 @@
...
 
 import java.util.List;
-import java.util.Map;
 
 /**
  * A basic http request abstraction. Http modules needs to implement this interface to integrate with the
  * server package's rest handling.
  */
-public interface HttpRequest {
+public interface HttpRequest extends BasicHttpRequest {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/http/HttpRequest.java:22]: Is this a general preference of yours or it's just this particular case?

I think the two options are:
 * an interface extends another and adds more methods; in case the base interface is passed around and the code has some specialized logic it can use `instance of` checks to downcast
 * have a single rich interface and throw "not implemented"

[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/http/HttpRequest.java:22]: FWIW the body needed for auditing authn success (which, again, is one of the things that's not dealt with properly by this PR) so pending other discussions, we might not need a request interface with a request body....

[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/http/HttpRequest.java:22]: > don't use anonymous implementations of BasicHttpRequest instead use a concrete HttpRequestWithoutBody.

That's separate from the interface conversation, I think. There's going to be some light wrapping needed because the request classes that netty uses are not available to the Security plugin module.

[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/http/HttpRequest.java:22]: > Is this a general preference of yours or it's just this particular case?

I think general ? ... basically I think the interface should represent the concept and any sub interfaces should be specializations of that concept. (IMO) Here the concept is a http request and not having access to the body the a specialization.  I understand the throwing not implemented from many methods is also it's own issue, so this is more subjective / preference than any rule.  However, I do think as-is this has some readability / expectation issues. Maybe just changing the names would help ... as-is there no indication to what the difference between a HttpRequest and BasicHttpRequest means...java doc can help but good names > doc.  I think i still prefer what I suggested above ...but having something like FullHttpRequest extends PartialHttpRequest makes more readily apparent what to expect from the contract.

[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/http/HttpRequest.java:22]: > ... has some specialized logic it can use instance of checks to downcast

I would strongly advise against a design that necessitates instanceof checks to downcast

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java (hunk 6 above)
Comment (jakelandis) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:78]: I think I remember you said that we need to associate this authenticated context with the request since the context will get lost since it may not part of the same call stack... but can you detail abit more why that is ? Isn't both this code (storing the authenticated context) and the code that restores the authenticated context happening on the same eventloop thread ?  (also, what thread does the current authentication happen on ->  the eventloop thread?)

Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:78]: > Isn't both this code (storing the authenticated context) and the code that restores the authenticated context happening on the same eventloop thread ?

This is a great point to raise!
Yes, that is correct, both pieces of code (setting in and then upacking) run under the same thread.
The fundamental issue is with the call stack. I'll explain briefly but this might be worth some sync conv.
The authenticated context needs to be set for a specific request only. This is crucial. The way we normally achieve this is to set it in a try-with-resource block and, just after that, inside the block, we call the handler for the request. When the handler for that request returns, the try-block closes, and the thread-context is restored to the un-authenticated state. Hence, there is no danger that a single handler being passed-in a single request, somehow starts executing another request.
Contrast this to setting the authenticated context earlier in the process, in the netty channel handler. The concept of a full request is not there yet, there are only request pieces. A good guess is to set an authenticated thread-context when the last piece of a HTTP request is forwarded. This relies on the fact that the same call stack that forwards the last piece is going to invoke the handler for the completely assembled request (I also tested it and works). The danger, of course, is that theoretically the full-request might be buffered somewhere in a downstream netty channel handler or ES code. In this case there might be multiple requests that are executed under a single authenticated thread-context.

Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:78]: > what thread does the existing (before this PR) authentication happen on -> the eventloop thread?

Yes the same eventloop thread does the authentication currently too.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HeaderValidationException.java [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+
+public class HeaderValidationException extends Exception {
+    private final boolean closeChannel;
+
+    public HeaderValidationException(ElasticsearchException cause, boolean closeChannel) {
+        super(cause);
+        this.closeChannel = closeChannel;
+    }
+
+    public boolean shouldCloseChannel() {
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HeaderValidationException.java:21]: It will indeed

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java (hunk 6 above)
Comment (jakelandis) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java:78]: > I'll explain briefly but this might be worth some sync conv.

Thanks, I scheduled a discussion on Monday.
